Polygon_mesh_processing/test/Polygon_mesh_processing/test_pmp_polyhedral_envelope.cpp

```diff
@@ -0,0 +1,120 @@
+    assert(envelope(sm));
+    assert(envelope(poly));
+    assert(envelope(points, triangles));
+
+    envelope=CGAL::Polyhedral_envelope<EPIC>(points,triangles,0.0001);
+    assert(!envelope.is_empty());
+    assert(envelope(sm));
+    assert(envelope(poly));
+    assert(envelope(points, triangles));
+
+    std::vector<decltype(sm)::Face_index> subfaces(faces(sm).begin(), std::next(faces(sm).begin(), 150));
+    envelope=CGAL::Polyhedral_envelope<EPIC>(subfaces,sm,0.0001);
+    assert(!envelope.is_empty());
+    assert(!envelope(sm));
+    assert(!envelope(poly));
+    assert(!envelope(points, triangles));
+  }
+
+  // build from different kernels
+  {
+    auto sm_np = CGAL::parameters::vertex_point_map(
+      CGAL::make_cartesian_converter_property_map<EPEC::Point_3>(
+        get(boost::vertex_point, sm)));
+    auto poly_np = CGAL::parameters::vertex_point_map(
+      CGAL::make_cartesian_converter_property_map<EPEC::Point_3>(
+        get(boost::vertex_point, poly)));
+    auto soup_np =  CGAL::parameters::point_map(
+      CGAL::make_cartesian_converter_property_map<EPEC::Point_3>(
+        CGAL::Identity_property_map<EPIC::Point_3>()));
+
+    CGAL::Polyhedral_envelope<EPEC> envelope(poly,0.0001,poly_np);
+    assert(envelope(sm, sm_np));
+    assert(envelope(poly, poly_np));
+    assert(envelope(points, triangles, soup_np));
+
+    envelope = CGAL::Polyhedral_envelope<EPEC>(points,triangles,0.0001, soup_np);
+    assert(envelope(sm, sm_np));
+    assert(envelope(poly, poly_np));
+    assert(envelope(points, triangles, soup_np));
+
+    std::vector<decltype(sm)::Face_index> subfaces(faces(sm).begin(), std::next(faces(sm).begin(), 150));
+    envelope=CGAL::Polyhedral_envelope<EPEC>(subfaces,sm,0.0001,sm_np);
+    assert(!envelope.is_empty());
+    assert(!envelope(sm, sm_np));
+    assert(!envelope(poly, poly_np));
+    assert(!envelope(points, triangles, soup_np));
+  }
+}
+
+void test_remove_si()
+{
+  std::cout << "---- test_remove_si() ----\n";
+  CGAL::Surface_mesh<EPIC::Point_3> tm;
+  std::ifstream in("data/pig.off");
+  in >> tm;
+  assert(tm.vertices().size()!=0);
+
+  // try with a small bound --> reject fix
+  PMP::experimental::remove_self_intersections(tm, CGAL::parameters::polyhedral_envelope_epsilon(0.0001));
+  assert(PMP::does_self_intersect(tm));
+  // try with a larger bound --> accept fix
+  PMP::experimental::remove_self_intersections(tm, CGAL::parameters::polyhedral_envelope_epsilon(0.001));
+  assert(!PMP::does_self_intersect(tm));
+}
+
+void cube_test()
+{
+  std::cout << "---- cube_test() ----\n";
+  CGAL::Surface_mesh<EPIC::Point_3> tm;
+  std::ifstream in("data-coref/cube_meshed.off");
+  in >> tm;
+  assert(tm.vertices().size()!=0);
+
+  typedef EPIC::Point_3 P;
+  CGAL::Polyhedral_envelope<EPIC> envelope(tm, 0.1);
+  assert(envelope(P(0,0,0), P(1,0,0), P(1,1,0)));
+  assert(envelope(P(0,0,0.05), P(1,0,0.05), P(1,1,0.05)));
+  // assert(envelope(P(0,0,0.1), P(1,0,0.1), P(1,1,0.1))); // NOT WORKING
```

Comment:
@afabri is it expected that this one is not working?

Reply:
I would say yes, as you have chosen `epsilon = 0.1`  which means `delta ~= 0.05`.  It is a double sided hull, that is lying inside the volume of the cube does not mean to be inside the hull of the surface of the cube.

Reply:
I am wondering if the `operator()`  is a good choice.   It might be better to have something like `contains()`.